tusclient/uploader.py

```diff
@@ -211,7 +223,35 @@ def upload_chunk(self):
         """
         Upload chunk of file.
         """
+        self._retried = 0
         self._do_request()
         self.offset = int(self.request.response_headers.get('upload-offset'))
         msg = '{} bytes uploaded ...'.format(self.offset)
         print(msg)
+
+    def _do_request(self):
+        # TODO: Maybe the request should not be re-created everytime.
+        #      The request handle could be left open until upload is done instead.
+        self.request = TusRequest(self)
+        try:
+            self.request.perform()
+            self.verify_upload()
+        except TusUploadFailed as error:
+            self.request.close()
+            self._retry_or_cry(error)
+        finally:
+            self.request.close()
+
+    def _retry_or_cry(self, error):
+        if self.retries > self._retried:
+            time.sleep(self.retry_delay)
```

Comment:
You can have a look at the Android example which uses the Java library. There you have a background thread (https://github.com/tus/tus-android-client/blob/master/example/src/main/java/io/tus/android/example/MainActivity.java#L165-L168) which uploads one chunk at a time and always checks afterwards whether the upload should be stopped using `isCancelled()`. This thread is cancelled when the Pause button on the UI is pressed (https://github.com/tus/tus-android-client/blob/master/example/src/main/java/io/tus/android/example/MainActivity.java#L65-L71 and https://github.com/tus/tus-android-client/blob/master/example/src/main/java/io/tus/android/example/MainActivity.java#L92-L94). The reason that this works is that we have two different threads running in parallel, one for the UI and one responsible for the upload loop.